README
NAME
    YATG - Fast SNMP data poller daemon, with storage and graphing

VERSION
    version 5.130600

DESCRIPTION
    YATG is a daemon (background process) which at intervals wakes up and
    polls network devices for SNMP data, and then stores or prints that
    data. In this distribution are also included examples for presenting
    simple CGI web pages with graphs.

    YATG is flexible, efficient and powerful. It can poll a large number of
    devices with thousands of ports in just a few seconds. The configuration
    is very simple, and the defaults sane (it's designed for sysadmins,
    after all).

    You can use YATG both for historical logging, such as traffic counters
    on ports, as well as short-term monitoring which might feed into, say,
    Nagios. Wherever possible, data is translated to human-friendly formats
    for storage, such as using Leaf Names instead of OIDs, translated values
    ("up", "down", etc) and device port names rather than SNMP Interface
    Indexes.

How Does It Work?
    At startup, "yatg_updater" loads its configuration from local files and
    a database, performs some basic SNMP connections to build a cache about
    device capabilities and so on, and then goes to sleep.

    Periodically, as determined by the configuration, "yatg_updater" wakes
    up and polls all devices, then stores results, again according to
    instructions in the configuration.

    If you have only the essential dependencies installed (see below) then
    you can only output results to STDOUT. With other modules, you have more
    options such as local or remote disk, or memcached based storage.

    "yatg_updater" will re-load all its configuration if given a HUP signal.
    If you run the daemon persistently (for example with "daemontools") then
    a cron job once a day is a good way to refresh the configuration. There
    is reference to this in one of the bundled example files.

What's in this distribution
    "yatg_updater"
        This is the main application, designed to be run persistently. It
        does not accept any input and only produces output when in debugging
        mode. It is a smart wrapper for the SNMP::Effective module.

    YATG::Store family of modules
        These are modules which take the SNMP poll results and store them to
        either local Disk, a Memcached server, the disk on a remote
        networked server, or Nagios via NSCA.

    YATG::Retrieve family of modules
        These are modules which read stored results back to you, for a given
        time window. The data can be retrieved from local Disk, a Memcached
        server, or the disk on a remote networked server.

    RPC::Serialized handlers
        If storing and/or retrieving on a remote networked server, it should
        run an instance of RPC::Serialized, and these are the RPC Handlers
        for that server (see that module's documentation for further
        details).

    CGI For the special case of viewing graphs of disk-based poll results
        for switch port traffic counters, there is are two CGI scripts. One
        is a wrapper which presents an HTML page embedded with PNG images
        created from the other script.

    Examples
        The "examples/" folder includes a copy of each of the files you
        should need for a complete deployment of YATG. Obviously some of
        them contain dummy data.

Where to go from here
    To begin with, you probably want to see how to configure "yatg_updater"
    in YATG::Config.

    Alongside that, there are examples of all the files you should need to
    install, in the "examples/" folder of this distribution.

    Each of the Store and Retrieve modules might have additional Perl module
    dependencies (i.e. from CPAN) - see the relevant docs for more details.

LOGGING and TESTING
    This module uses "Log::Dispatch::Syslog" for logging, and by default
    will log timing data to your system's syslog service. More information
    is provided in the YATG::Config documentation.

    To run in debug mode, where timing data is output to standard out rather
    than syslog, set the environment variable "YATG_DEBUG" to a true value.

    To run the poller just once, set the "YATG_SINGLE_RUN" environment
    variable to a true value. This is great for development. It makes
    "yatg_updater" load its configuration, generate the device hints cache,
    sleep and then run just one poll cycle before exiting.

    To override the interval between polling runs, set the "YATG_INTERVAL"
    environment variable to a number of seconds.

    For example:

     YATG_DEBUG=1 YATG_SINGLE_RUN=1 /usr/bin/yatg_updater /etc/yatg.yml

SEE ALSO
    SNMP::Effective
        This system uses SNMP::Effective at its core do the polling.

    RPC::Serialized
        Store polled data on another server using RPC::Serialized.

AUTHOR
    Oliver Gorwits <oliver@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by University of Oxford.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
5.130600  2013-03-01 16:45:46 Europe/London

  [NEW FEATURES]

  * newhosts_watch option, which triggers host-list update if modified

5.130520  2013-02-21 10:45:40 Europe/London

  [ENHANCEMENTS]

  * Add threshold check to Errors and Discards NSCA reporter - min 1/minute

  [BUG FIXES]

  * POD fixes

5.130511  2013-02-20 18:03:29 Europe/London

  [BUG FIXES]

  * Fix bug with multiple community strings - only first was used
  * Fix check of Dialer interface status which can be "dormant" (~~ "up")

5.130510_001 2013-02-20 13:04:42 Europe/London

  [NEW FEATURES]

  * New environment variable YATG_INTERVAL to override interval config val
  * NSCA plugin now has per-leaf reports for ifOperStatus, ifInErrors, ifInDiscards
  * New config option dbi_host_query to load hostnames for IPs (used in NSCA plugin only)
  * New config option dbi_interfaces_query to filter interfaces repoted by NSCA plugin
  * New ignore_*_descr config to filter interfaces by description in NSCA plugin

  [ENHANCEMENTS]

  * Remove all non-essential dependencies. User must install deps for each plugin
  * NSCA plugin no longer tries DNS PTR lookup - host names must be suitable for Nagios
  * NSCA plugin no longer searches for send_nsca in PATH
  * Use Data::Printer instead of Data::Dumper in STDOUT backend module

  [BUG FIXES]

  * Add explicit dependencies on YAML::XS and Sys::Syslog 0.25

4.112532  2011-09-10 14:24:35 Europe/London

  * Make NSCA config optional (D. Walrond)
  * Remove VERSION so that release is indexed properly

4.112530  2011-09-10 12:42:15 Europe/London

  * Port to Dist::Zilla

yatg (3.02) UNRELEASED; urgency=low

  * alter ignored interfaces list when storing data not to exclude port-channels

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Wed, 02 Jun 2010 11:51:01 +0100

yatg (3.01) UNRELEASED; urgency=low

  * yatg_updater now retrieves (new) list of devices on each poll run
  * can now have multiple store backends per OID
  * new store backend NSCA for submitting passive Nagios service check results

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Sun, 16 Aug 2009 11:24:09 +0100

yatg (2.0.1-1) UNRELEASED; urgency=low

  * Move to Module::Install
  * Sanitize the CGI

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Sat,  8 Nov 2008 18:34:00 +0000

yatg (2.0.0-1) unstable; urgency=low

  * Use gmtime rather than localtime.

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Wed,  9 Jan 2008 09:35:58 +0000

yatg (1.3.3-1) stable; urgency=low

  * Added Tutorial POD from Nelson Ferraz

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Fri, 30 Nov 2007 15:34:00 +0000

yatg (1.3.2-1) stable; urgency=low

  * Missed file from MANIFEST

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Thu, 11 Oct 2007 14:22:00 +0100

yatg (1.3.1-1) stable; urgency=low

  * Documentation fixes
  * Include nagios check example script

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Tue,  9 Oct 2007 10:54:00 +0100

yatg (1.3-1) stable; urgency=low

  * Update for SNMP::Effective 0.4 (which is newer than 1.0)
  * Documentation
  * Perl packaging

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Sun,  7 Oct 2007 22:15:00 +0100

yatg (1.2-2) stable; urgency=low

  * Add a cheat for adding multiple data sources (load balanced ports)
  * Check version of Module::Multiconf so that no_validation is supported

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Wed,  3 Oct 2007 23:33:00 +0100

yatg (1.2-1) unstable; urgency=low

  * Disable config validation from CGI - parts might be missing
  * Fix for initial values of Diff data
  * Fix for various URI escaping oopsies in yatgview.cgi
  * Add check for SNMP::Effective version because its API is going to change
  * Add copyright notice to all files

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Fri, 21 Sep 2007 12:38:00 +0100

yatg (1.1-2) unstable; urgency=low

  * Update ignore port type list to include "netflow"

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Thu, 20 Sep 2007 14:16:00 +0100

yatg (1.1-1) unstable; urgency=low

  * Add storage of device list and port list to Memcached

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Wed, 19 Sep 2007 22:37:00 +0100

yatg (1.0-1) unstable; urgency=low

  * Major bump to newly engineered version

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Wed, 19 Sep 2007 19:19:00 +0100



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Fast SNMP data poller daemon, with storage and graphing",
  "AUTHOR" => "Oliver Gorwits <oliver\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "YATG",
  "EXE_FILES" => [
    "bin/yatg_updater"
  ],
  "LICENSE" => "perl",
  "NAME" => "YATG",
  "PREREQ_PM" => {
    "Class::Data::Inheritable" => 0,
    "DBI" => 0,
    "Data::Printer" => 0,
    "Exporter" => 0,
    "File::Basename" => 0,
    "File::Spec::Functions" => 0,
    "File::stat" => 0,
    "FindBin" => 0,
    "Log::Dispatch::Syslog" => 0,
    "Module::MultiConf" => "0.0301",
    "Net::Netmask" => 0,
    "POSIX" => 0,
    "Readonly" => 0,
    "Regexp::Common" => 0,
    "SNMP" => 0,
    "SNMP::Effective" => 0,
    "Symbol" => 0,
    "Sys::Syslog" => "0.25",
    "Time::HiRes" => 0,
    "YAML::XS" => 0,
    "base" => 0,
    "strict" => 0,
    "vars" => 0,
    "warnings" => 0
  },
  "VERSION" => "5.130600",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



